index.js

```diff
@@ -0,0 +1,346 @@
+
+        fs.access(dbDir, fs.constants.W_OK, (err) => {
+          if (err && err.code === 'ENOENT') {
+            const msg = `the directory ${dbDir} does not exist, please create`
+
+            return next(new Error(msg))
+          }
+          if (err) {
+            return cb(err)
+          }
+
+          this._initDb(next)
+        })
+      }
+    ], cb)
+  }
+
+  asyncQuery (args) {
+    if (this.opts.isNotWorkerSpawned) {
+      throw new Error('ERR_WORKER_HAS_NOT_BEEN_SPAWNED')
+    }
+
+    const { action, sql, params } = { ...args }
+
+    return new Promise((resolve, reject) => {
+      const workerArr = [...this._workers]
+        .reverse()
+        .find(([, jobData]) => {
+          return !jobData.job && jobData.isOnline
+        })
+      const job = {
+        resolve,
+        reject,
+        message: { action, sql, params }
+      }
+
+      if (!Array.isArray(workerArr)) {
+        this._queue.push(job)
+
+        return
+      }
+
+      const [worker, jobData] = workerArr
+
+      jobData.job = job
+      worker.postMessage(job.message)
+    })
+  }
+
+  initializeWalCheckpointRestart (ms = 10000) {
+    const {
+      dbPath,
+      unacceptableWalFileSize,
+      isNotWorkerSpawned
+    } = this.opts
+    const walFile = `${dbPath}-wal`
+
+    clearInterval(this._checkpointRestartInterval)
+
+    this._checkpointRestartInterval = setInterval(() => {
+      fs.stat(walFile, (err, stat) => {
+        if (err) {
+          if (err.code === 'ENOENT') return
+
+          throw err
+        }
+        if (stat.size < unacceptableWalFileSize) {
+          return
+        }
+        if (isNotWorkerSpawned) {
+          this.db.pragma('wal_checkpoint(RESTART)')
+
+          return
+        }
+
+        this.asyncQuery({
+          action: DB_WORKER_ACTIONS.EXEC_PRAGMA,
+          sql: 'wal_checkpoint(RESTART)'
+        }).catch((err) => console.error(err))
+      })
+    }, ms).unref()
+  }
+
+  _getRequiredWorkersCount () {
+    const cpusCount = os.cpus().length
+    const minRequiredWorkersCount = Math.max(
+      cpusCount,
+      this.opts.minWorkersCount
+    )
+    const maxRequiredWorkersCount = Math.min(
+      minRequiredWorkersCount,
+      this.opts.maxWorkersCount
+    )
+
+    return Math.max(maxRequiredWorkersCount, 1)
+  }
+
+  _spawnWorkers (workerPath, workerData) {
+    const spawn = () => {
+      const worker = new Worker(
+        workerPath,
+        { workerData }
+      )
+      const jobData = {
+        job: null,
+        error: null,
+        timer: null,
+        isOnline: false
+      }
+      this._workers.set(worker, jobData)
+
+      const poll = () => {
+        jobData.isOnline = true
+
+        if (
+          this._queue.length !== 0 &&
+          !jobData.job
+        ) {
+          jobData.job = this._queue.shift()
+          worker.postMessage(jobData.job.message)
+
+          return
+        }
+
+        jobData.timer = setTimeout(poll, 3000)
```

Comment:
maybe polling just every three seconds is not enough when there is a full queue with 100 jobs?

Reply:
The 3 seconds rule is just for when there are no jobs in queue

Reply:
So we are pulling a job by a worker when adding a task by `asyncQuery` method or when a worker will finish a current taken task `.on('message', process)` or as Ezequiel said